### DISCUSSION TITLE
Bardzo niska wydajność UNCACHED dysków w wirtualkach bazujących na procesorach Intel.

### DISCUSSION THREAD

**Discussion by plukawski** [AzureCommunityPL/group#193] [Pomoc]
Witam wszystkich,
ostatnio odkryłem dziwną anomalię w Azure. Podczas testowania wirtualnych maszyn pod obciążenia typu webapi oraz bazy SQL zauważyłem, że maszyny bazujące na procesorach Intel mają drastycznie gorszą wydajność UNCACHED dysków niż ich odpowiedniki bazujące na procesorach AMD. Na przykład import 10 000 rekordów do bazy Sqlite3 na maszynie **D4ds_v4** trwa ok 10 minut, gdzie na maszynie **D4as_v4** ten sam import trwa ok 180 sekund. Wyniki są powtarzalne. Obie maszyny są w tym samym regionie: West Europe. Obie maszyny mają ten sam typ dysku podpięty (Permium SSD, MAX IOPS 1100, Max Throughput 125). System operacyjny jest też ten sam: Windows 10 1909 z najnowszymi aktualizacjami.

Różnice w wydajności są bardzo duże i w żaden sposób nie wynikają z oficjalnej dookumentacji, według której ich w ogóle nie powinno być. Czy ktoś mógłby powiedzieć, czy spotkał się z czymś podobnym i co może być tego przyczyną? To samo zauważyłem porównując maszyny **E4as_v4** oraz **E4ds_v4**.

**Comment by kamil-mrzyglod** [comment]
Patrząc po ACU to jednak maszyny oparte o AMD są mocniejsze od tych z Intelem. Zawsze możesz założyć ticket do supportu aby pomogli Ci to przeanalizowac krok po kroku i znaleźć wąskie gardło. Patrzyłeś na performance, to na pewno procek jest tak mocno obciążany?

**Reply by plukawski** [reply]
Właśnie to nie procek sam w sobie jest problemem. Jak sobie patrzyłem na wydajność procesorów, to była bardzo zbliżona. Zresztą w moim teście procesor nie był nawet w połowie obciążony. Podsystem IO na Intelu siada i jest to dla mnie dziwne, bo różnica w jego wydajności jest ogromna. Na tyle ogromna, że dyskwalifikuje postawienie jakiejkolwiek bazy na takich wirtualnych maszynach w moim przypadku.

**Comment by kaluzaaa** [comment]
@plukawski testowałeś na jednej maszynie czy na kilku?

**Reply by plukawski** [reply]
@kaluzaaa narazie przetestowałem na dwóch parach (tych co wspomniałem wyżej). Testy wykonywałęm kilkukrotnie. Nawet uruchamiałem ponownie maszyny, dealokowałem i alokowałem na nowo. Za każdym razem wyniki były podobne, w sensie na tych z Intelem wydajność UNCACHED dysku była tragiczna.

**Comment by smereczynski** [comment]
Porównując sobie te procesory (może nie identyczny Intel, ale praktycznie to samo), widać duże różnice, które mogą mieć znaczenie: https://www.cpu-monkey.com/en/compare_cpu-amd_epyc_7452-992-vs-intel_xeon_platinum_8276m-1062

Kolosalna różnica w pamięci L3, szybsza pamięć, nowsza szyna PCI-E i więcej linii tej szyny. Nie wiem, czy to może być powód tak znacznej różnicy, ale z pewnością nie jest to bez wpływu. Maszyny z procesorami AMD nie od dziś są uważane, za lepiej sprawujące się w tematach obciążających storage. Oczywiście architektura Azure to nie to samo, co utylizacja bare metal, ale na moje oko Twój przypadek to pochodna tego stanu rzeczy.

**Reply by plukawski** [reply]
@smereczynski @kamil-mrzyglod Przepraszam, że nie odpowiadałem przez dłuższy czas, ale spędziłem trochę czasu nad analizą tego problemu. Otóż zachowanie Azure jest bardzo dziwne. Jak stworzyłem nową maszynę wirtualną **D4ds_v4** od zera z dyskiem Premium SSD P15, to czas importu 10 000 rekordów był na niej tylko o minutę dłuższy niż na maszynie D4as_v4 (trwał niecałe 240 sekund, czyli niecałe 4 minuty). Wynik był powtarzalny. Niestety ta sama operacja na starej maszynie **D4ds_v4** nadal trwała prawie 10 minut (wynik powtarzalny).
I teraz najciekawsza część - jak na tej nowej maszynie zmieniłem dysk na Standard SSD E10, a później przywróciłem spowrotem na Premium SSD P15 to problem z długim czasem importu pojawił się - teraz import też na tej nowej trwa ok 10 minut, czyli teraz zarówno nowa maszyna jak i stara importują tak samo wolno - ok 10 minut. Co ciekawe syntetyczne benchmarki nie wykazują tych różnic. Tylko obciążenia typowo bazodanowe je wykazują, bo wymuszają zapisy niebuforowe.

Tego problemu wydajnościowego nie jestem w stanie zreprodukować na maszynach opartych o AMD EPYC. Mi się wydaje, że przyczyną tego stanu ma dwa źródła:
1. Procesory Intel Xeon faktycznie mają niższą wydajność storage'u, ale nie jest to różnica aż tak drastyczna (ok 30%)
2. W konfiguracji podsystemu IO dla maszyn opartych o procesory Intel jest jakiś bug po stronie Azure, który możę powodować spadek wydajności UNCACHED nawet kilkakrotnie.
